File: 165-compare-version-numbers/compare-version-numbers.go
import (
    "strconv"
    "strings"
)

func compareVersion(version1 string, version2 string) int {
    parts1 := strings.Split(version1, ".")
    parts2 := strings.Split(version2, ".")
    
    maxLen := len(parts1)
    if len(parts2) > maxLen {
        maxLen = len(parts2)
    }
    
    for i := 0; i < maxLen; i++ {
        var v1, v2 int
        
        if i < len(parts1) {
            v1, _ = strconv.Atoi(parts1[i])
        }
        
        if i < len(parts2) {
            v2, _ = strconv.Atoi(parts2[i])
        }
        
        if v1 < v2 {
            return -1
        } else if v1 > v2 {
            return 1
        }
    }
    
    return 0
}

/*
 * Time Complexity: O(n + m)
 * Space Complexity: O(n + m) for split approach, O(1) for parsing
 */

File: 165-compare-version-numbers/compare-version-numbers.rs
impl Solution {
    pub fn compare_version(version1: String, version2: String) -> i32 {
        let parts1: Vec<&str> = version1.split('.').collect();
        let parts2: Vec<&str> = version2.split('.').collect();
        
        let max_len = parts1.len().max(parts2.len());
        
        for i in 0..max_len {
            let v1 = if i < parts1.len() {
                parts1[i].parse::<i32>().unwrap_or(0)
            } else {
                0
            };
            
            let v2 = if i < parts2.len() {
                parts2[i].parse::<i32>().unwrap_or(0)
            } else {
                0
            };
            
            match v1.cmp(&v2) {
                std::cmp::Ordering::Less => return -1,
                std::cmp::Ordering::Greater => return 1,
                std::cmp::Ordering::Equal => continue,
            }
        }
        
        0
    }
}

/*
 * TC: O(n + m) for parsing and comparison
 * SC: O(n + m) for storing split parts
 */